feat: handle various month and year durations

RonMcKay opened the pull request:
This PR fixes some problems with durations and step sizes being specified as number of days. This leads to problems in leap year or months with different number of days e.g.

This implementation improves on this by utilizing `dateutil.relativedelta.relativedelta`, which handles such corner cases nicely. Additionally the maximum value for `max`, `inf` and `infinite` for the duration value is getting set by using `datetime.MAXYEAR`. This was necessary as adding `relativedelta` objects can lead to `OverflowError`.

RonMcKay committed:
Implement `duration` and `step` as relativedelta

By using `dateutil.relativedelta.relativedelta` instances it is possible
to easily model durations and step values of 'month' and 'year' with
varying lengths (such as in a leap year or from February to March e.g.).

RonMcKay committed:
Set maximum possible date

'inf', 'infinity' and 'max' had a too high number of days assigned which
causes `OverflowError` on datetime objects while using relativedelta
instances.

This commit catches the `OverflowError` and sets the duration to the
number of days between today and the maximum possible date.

Akuukis commented:
Thanks!

Apologies for master branch - it turns out I've forgot to push last changes. You should be able to rebase without conflicts.

I tried out the PR on my files and got plenty of errors like these (shortlist from 500+ lines):
```
WARNING: Using defaults, because cannot parse mark "3": unsupported operand type(s) for +: 'datetime.date' and 'int'
WARNING: Using defaults, because cannot parse mark "3": unsupported operand type(s) for +: 'datetime.date' and 'int'
WARNING: Using defaults, because cannot parse mark "5@2021-04-01": unsupported operand type(s) for +: 'datetime.date' and 'int'
WARNING: Using defaults, because cannot parse mark "7": unsupported operand type(s) for +: 'datetime.date' and 'int'
WARNING: Using defaults, because cannot parse mark "30": unsupported operand type(s) for +: 'datetime.date' and 'int'
WARNING: Using defaults, because cannot parse mark "720": unsupported operand type(s) for +: 'datetime.date' and 'int'
WARNING: Using defaults, because cannot parse mark "60 @ 2021-11-01": unsupported operand type(s) for +: 'datetime.date' and 'int'
WARNING: Using defaults, because cannot parse mark "1@2022-03-03": unsupported operand type(s) for +: 'datetime.date' and 'int'
```

Looks like numeric length is not handled anymore. See commit `ab05d0b`, I added a regression test for this.

Akuukis reviewed (changes requested):
^

RonMcKay commented:
Ah, did not know that just a numeric specification should also be supported. There was nothing in the README mentioning this. I will rebase and fix that.

RonMcKay committed:
Allow numeric length with default unit 'day'

Akuukis commented on beancount_interpolate/recur.py:
```diff
@@ -1,34 +1,33 @@
-    period = math.floor( duration / step )
+    period = get_number_of_txn(begin_date, duration, step, config['max_new_tx'])
 
     if(period > config['max_new_tx']):
         period = config['max_new_tx']
         duration = period * step
```
this block seems redundant now, as you're handling `max_new_tx` within `get_number_of_txn`.

Akuukis commented on beancount_interpolate/common.py:
```diff
@@ -103,26 +126,26 @@ def distribute_over_period(params, default_date, total_value, config):
-    period = math.floor( duration / step )
+    period = get_number_of_txn(begin_date, duration, step, max_txn=config['max_new_tx'])
 
     if(period > config['max_new_tx']):
         period = config['max_new_tx']
         duration = period * step
```
this block seems redundant now, as you're handling max_new_tx within get_number_of_txn.

Akuukis commented on beancount_interpolate/common.py:
```diff
@@ -26,6 +26,24 @@ def extract_mark_posting(posting, config):
     return False
 
 
+def get_number_of_txn(begin_date, duration, step, max_txn=float('inf')):
+    """
+    Computes the number of transactions within a given interval and step length.
+
+    Note: As the current implementation is a bit costly with a very large duration
+    and a small step size there is a `max_txn` parameter which can be set.
+    This is meant to be used with the config `max_new_txn`.
+    Upon exceeding `max_txn` the function returns `max_txn + 1`.
+    """
+    n_txn = 0
+    end_date = begin_date + duration
+    while begin_date + n_txn * step < end_date:
+        n_txn += 1
+        if n_txn > max_txn:
+            break
+    return n_txn
```
Yup, it's ~50% costlier overall for my ledger.

I timed my ledger 3 times (`$ time bean-check --no-cache ledger/main.bean`):
- plugin disabled: 1.5-1.6 seconds
- master: 16.5-17.0 seconds
- PR: 23.2 - 23.9 seconds

I'm not familiar with `relativedelta`, but does it have a equilavent function/method for manipulating durations and expressing then in various units, like for example [momentjs has with `asMonths` and similar methods](https://momentjs.com/docs/#/durations/)?

RonMcKay replied:
see [my comment](https://github.com/Akuukis/beancount_interpolate/pull/19#issuecomment-1073299033)

Akuukis reviewed (changes requested):
Overall it works good :+1: See comments inline.

Little concern is that I had few new balance errors in the ledger, but they were only on few already otherwise questionable accounts and accounting practices, so that's fine.

Indeed, README doesn't mention plain numbers. Do you agree with a paragraph like this?

```markdown
The content of meta tag consists of 4 parts, every of them optional:
1. amount, a number. Defaults to 1.
2. period, a case-insensitive keyword "day", "week", "month" or "year". Defaults to "day".
3. start date, prefixed with `@` sign. Defaults to transaction date.
4. step size, prefixed with `/` sign. Can be either keyword from as period. Defaults to "day".

Summary: `[x] [period] [@ YYYY-MM[-DD]] [/ step]`
```

RonMcKay committed:
Improve performance of `get_number_of_txn`

RonMcKay commented:
> Overall it works good 👍 See comments inline.
> 
> Little concern is that I had few new balance errors in the ledger, but they were only on few already otherwise questionable accounts and accounting practices, so that's fine.
> 
> Indeed, README doesn't mention plain numbers. Do you agree with a paragraph like this?
> 
> ```gfm
> The content of meta tag consists of 4 parts, every of them optional:
> 1. amount, a number. Defaults to 1.
> 2. period, a case-insensitive keyword "day", "week", "month" or "year". Defaults to "day".
> 3. start date, prefixed with `@` sign. Defaults to transaction date.
> 4. step size, prefixed with `/` sign. Can be either keyword from as period. Defaults to "day".
> 
> Summary: `[x] [period] [@ YYYY-MM[-DD]] [/ step]`
> ```

This is currently not correct as one can not combine `[x]` and `[period]`. It rather has to be `[x|period]` as in the code above `RE_PARSING`. When I open the second PR where I implemented the more flexible duration / step then we can adapt the README like that.

I pushed another commit where I improved the performance of the `get_number_of_txn` function by utilizing that we only have unique units as a step size. Can you try that out on your ledger? Should be way faster than before. This change also answers the comments regarding `if(period > config['max_new_tx'])` as the function itself does no longer handle this and the `if` condition can stay.

Akuukis commented:
Right, I fixed the readme and pushed to master. As part of the next PR, please adjust it back.

Code looks good. I timed it again 3 times and it's faster but anyways slow at 21.4 to 21.9 seconds, that's still an ~40% increase.

I poked around with some experiments, and found put that this change on top of your PR makes everything fast again at 16.5 to 17.3 seconds.

![image](https://user-images.githubusercontent.com/7459419/159219538-79d77443-a415-41af-bf3b-52936c36203a.png)

RonMcKay commented:
ah I see, thanks for investigating. The recomputing of the `end_date` and the step in the `while` condition is indeed a waste of time by me.
The reason why I recomputed the `date` everytime starting from `begin_date` was the following:

Here we can see that starting at 31.01.2022 and going two months forward by calling `begin_date + 2 * rd` results in the 31.03.2022.
```
>>> from dateutil.relativedelta import relativedelta
>>> import datetime
>>> begin_date = datetime.datetime(2022, 1, 31).date()
>>> begin_date
datetime.date(2022, 1, 31)
>>> rd = relativedelta(months=+1)
>>> begin_date + 2 * rd
datetime.date(2022, 3, 31)
```

But if we instead overwrite a date, we get the 28.03.2022.
```
>>> date = begin_date + rd
>>> date
datetime.date(2022, 2, 28)
>>> date = date + rd
>>> date
datetime.date(2022, 3, 28)
```

I currently do not have an idea if this can cause any problems but if this speeds up the code significantly then we could leave it like that.

Could you test the speed of the following code snippet:

```
i = 0
end_date = begin_date + duration
tmp_date = begin_date + i * step
while tmp_date < end_date and tmp_date <= datetime.date.today():
    accumulated_remainder += total_value / period
    if(abs(round_to(accumulated_remainder)) >= abs(round_to(config['min_value']))):
        amount = D(str(round_to(accumulated_remainder)))
        accumulated_remainder -= amount
        amounts.append(amount)
        dates.append(tmp_date)
    i += 1
    tmp_date = begin_date + i * step
    if(tmp_date > datetime.date.today()):
        break
```

If the speed of this is acceptable then it could be a compromise between the two.

Akuukis commented:
Tested, this is as fast as previously :+1: 

Unrelated note, but this seems as a duplicate for `and tmp_date <= datetime.date.today():`, please remove it along the way:
```python
    if(tmp_date > datetime.date.today()):
        break
```

RonMcKay committed:
Improve performance of `distribute_over_period`

RonMcKay commented:
I added the commit.

Additionally I replaced `tmp_date <= datetime.date.today()` by `tmp_date <= today_date` and defined `today_date` before the loop. This might save a few ms extra.

Akuukis commented:
Thank you! Renamed & squashed.
